### PR TITLE
3D Tiles - Add allVisibleTilesLoaded event to Cesium3DTileset

### PR DESCRIPTION
**Pull request by lilleyse**
For #4400
From: https://github.com/AnalyticalGraphicsInc/cesium/pull/4401#issuecomment-251687673

The `loadedProgress` event doesn't take attempted requests into account, like when a request is made but the request scheduler is full, so checking when a tileset is loaded by checking that number processing and number of active requests are 0 may be faulty.

This `allVisibleTilesLoaded` event accounts for attempted requests and active requests / processing.

@mramato let me know how this works.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Code looks OK at a glance.  Why not rename to `allTilesLoaded`?  "Visible" is likely implicit to the user.

@mramato please test and merge.

**Comment by mramato** [comment]
I like the idea of `viewComplete` since we can re-use that name for imagery, terrain, and anywhere else people want to know if the scene is complete or not.  Also, I would prefer a boolean to an event, since my use case is to check in the postRender loop instead of having to subscribe to every tileset I create.

I'm checking out the behavior now.

**Comment by mramato** [comment]
Other than the desire for checkable flag, this resolves the original problem I was seeing.  Thanks @lilleyse!

**Comment by lilleyse** [comment]
I added a `viewComplete` flag to `Cesium3DTileset`. In order to make it work correctly on the first frame I changed the load process slightly so that the tileset isn't considered ready before its first update. Now the load state and ready look more like `Model`.

**Comment by mramato** [comment]
Given your latest changes, do you need @pjcozzi to look at this again before I test and merge?

**Comment by lilleyse** [comment]
Yes that's a good idea. Feel free to test though.

File: Source/Scene/Cesium3DTileset.js [hunk 1]
```diff
@@ -472,6 +479,23 @@ define([
...
+         * When <code>true</code>, all tiles in view are are loaded.
+         *
+         * @memberof Cesium3DTileset.prototype
+         *
+         * @type {Boolean}
+         * @readonly
+         *
+         * @default false
+         */
+        viewComplete : {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Can we please name this differently since it mixes poorly with the camera view semantics.  Why not stay consistent with the well understood tiles loaded paradigm?

File: Source/Scene/Cesium3DTileset.js [hunk 2]
```diff
@@ -472,6 +479,23 @@ define([
         },
 
         /**
+         * When <code>true</code>, all tiles in view are are loaded.
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Same comment.

Probably have a `@see` for each as well.

Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: > Probably have a @see for each as well.

Unless they are right next to each other in the doc like `ready` and `readyPromise`.

File: Source/Scene/Cesium3DTileset.js [hunk 3]
```diff
@@ -1468,8 +1527,23 @@ define([
      * @exception {DeveloperError} The tileset must be 3D Tiles version 0.0.  See https://github.com/AnalyticalGraphicsInc/3d-tiles#spec-status
      */
     Cesium3DTileset.prototype.update = function(frameState) {
+        if (!this._ready) {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Is all this really needed?  Couldn't we just have `_ready` and move some logic into `frameState.afterRender` to preserve the logic of the first `update`?

Either way, there should be a comment so we don't try to remove the "needless" complexity later.

Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Yeah looking at it now, it's more complicated than it needs to be. It's updated now.

File: Source/Scene/Cesium3DTileset.js (hunk 1 above)
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: For sure.

File: Source/Scene/Cesium3DTileset.js (hunk 3 above)
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Looks nice.

**Comment by lilleyse** [comment]
Updated with the name change.

**Comment by pjcozzi** [comment]
Thanks for the updates, @lilleyse!

@mramato can you retest with your scenario and merge when ready?

**Comment by pjcozzi** [comment]
@mramato any update?

**Comment by mramato** [comment]
Sorry, will get to this soon, tomorrow at the latest.

**Comment by mramato** [comment]
Just that one comment, this solves the original problem I had.  Thanks @lilleyse!

File: Source/Scene/Cesium3DTileset.js [hunk 4]
```diff
@@ -335,32 +355,32 @@ define([
...
 
         var that = this;
-
         this.loadTileset(tilesetUrl).then(function(data) {
             var tilesetJson = data.tilesetJson;
             that._asset = tilesetJson.asset;
             that._properties = tilesetJson.properties;
             that._geometricError = tilesetJson.geometricError;
             that._root = data.root;
-            that._readyPromise.resolve(that);
```
Comment (mramato) [inline Source/Scene/Cesium3DTileset.js]: @lilleyse any update here?  If resolving inside of the update is important, another solution would be to make sure we only reject inside the update as well, so it's at least consistent.  But I still think having to add it to the scene to have the promise resolve is odd (if not simply inconsistent with imagery)

Comment (mramato) [inline Source/Scene/Cesium3DTileset.js]: @lilleyse bump

Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: I haven't forgot about this yet - I'll try to get to it soon.

**Comment by pjcozzi** [comment]
@lilleyse is it trivial to finish and merge this?

**Comment by lilleyse** [comment]
Updated.

@mramato The ready promise behavior should be back to normal now.

There is one potential side effect - `tilesVisible` will be true between when the ready promise resolves and the first update occurs. Will this be a problem? It can be solved by adding a variable that marks when update is called for the first time, but I didn't want to add that clutter unless needed.

At least, the `allTilesLoaded` event should work as expected and only fire after the first update.

**Comment by lilleyse** [comment]
> There is one potential side effect - tilesVisible will be true between when the ready promise resolves and the first update occurs. Will this be a problem? It can be solved by adding a variable that marks when update is called for the first time, but I didn't want to add that clutter unless needed.

Fixes this and made the tests better at checking these corner cases. Ready now.

**Comment by pjcozzi** [comment]
LGTM
